### ISSUE TITLE
How to use slf4j-mock with Logback?

### ISSUE THREAD

**Issue by sebastiantf** [s4u/slf4j-mock#90]
I'm working on a **Spring Boot** project and I used _slf4j-mock_ to `Mockito.verify()` appropriate messages are logged.

Since there's **Logback** being pulled in by `spring-boot-starter-logging`, I excluded `logback-classic` in the `testCompile` scope in the `build.gradle`. So now I wasn't getting the **multiple SLF4J binding** errors and all my tests for the @Service classes were working fine.

But when I started writing tests for my @Controller classes using `MockMvc`,I realized it needs Logback for logging. Since I excluded Logback in `testCompile`, the controller tests wont even run, giving `NoClassDefFoundError` for classes in `logback-classic`.

How can I have both Logback and slf4j-mock together so that my @Controller tests and @Service tests run fine together?

**Comment by slawekjaranowski** [comment]
As we know `slf4j` support only one binding on classpath http://www.slf4j.org/manual.html#swapping.

This library provide additional binding which provide `slf4j` loggers with mock extension. 
So I don't see possibility to use `slf4j-mock` and another binding in one module - it depends on `slf4j` architekture.

One workaround can be create separate project module which only contains test classes for logging behavior and another modules for rest of application. In this case we can have different classpaths for tests in separate module.

Maybe someone has another idea?

**Comment by sebastiantf** [comment]
Thank you for the response. Appreciate it.

But I'm not sure I fully understand the workaround that you mentioned. The logging behaviour is part of the controller class itself, which also have some logic being executed. Shouldn't I test both the logic and the logging behaviour in the same test for that controller class?

Or did you mean to say to have two different modules with tests that test the same controller class, but one module tests only the logging behaviour and the other module tests only the logic?

**Comment by sebastiantf** [comment]
Also, the controller has logging behaviour that I have to test, for which I'd have to use `slf4j-mock`. But `MockMvc` will also try to initialize Logback for its own use. So I'm not sure splitting the test classes to modules would work around this problem.

**Comment by slawekjaranowski** [comment]
I looked some at logback, I found something like `Context Selectors` - https://logback.qos.ch/manual/loggingSeparation.html maybe it will be direction to provide own `ContextSelector` which will be responsible for manage logers mocks for logback.

**Comment by sebastiantf** [comment]
I think I figured out the issue. Even though I had excluded `logback-classic` from `testCompile`, there were explicit dependencies declared for `logback-core` and `logback-access`. I believe, them being in the classpath was the reason why `MockMvc` was still going for Logback.

So I now exclude `group: 'ch.qos.logback'` from `testCompile` so that all logback packages are excluded. The controller tests run fine now.

Thank you for spending time on this. I'm gonna close this issue now.
